internal/run/convert.go

```diff
@@ -0,0 +1,190 @@
+package run
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/helm/helm-2to3/pkg/v3"
+	"github.com/helm/helm-2to3/pkg/common"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	utils "github.com/maorfr/helm-plugin-utils/pkg"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/pelotech/drone-helm3/internal/env"
+)
+
+type Convert struct {
+	*config
+	release string
+
+	deleteV2Releases bool
+	dryRun bool
+	kubeContext string
+	kubeConfig string
+	label string
+	releaseVersionsMax int
+	tillerNS string
+
+	cmd cmd
+}
+
+func NewConvert(cfg env.Config, tillerNS string, kubeConfig string) *Convert {
+	return &Convert{
+		config: newConfig(cfg),
+		release: cfg.Release,
+		deleteV2Releases: cfg.DeleteV2Releases,
+		dryRun: cfg.DryRun,
+		kubeContext: cfg.Context,
+		kubeConfig: kubeConfig,
+		label: cfg.TillerLabel,
+		releaseVersionsMax: cfg.ReleaseVersionsMax,
+		tillerNS: tillerNS,
+	}
+}
+
+// Execute runs the `2to3 convert` command from Prepare
+// it detects if a conversion is needed and if not it will return no error
+// if a conversion is required then any returned error is related to a step
+// in the conversion process
+func (c *Convert) Execute() error {
+
+	clientSet, err := clientSetFromKubeConfig(c.kubeConfig)
+	if err != nil {
+		return err
+	}
+
+	// If Tiller is not present, utils.ListReleasesWithKubeConfig will panic
+	if !isTillerPresent(clientSet, "name=tiller,app=helm") {
+		log.Println("There is no Tiller Deployment running")
+		return nil
+	}
+
+	// Check for existence of v2 Release (only if Tiller is running)
+	var v2ReleaseFound bool
+
+	listOptions := utils.ListOptions{
+		ReleaseName: c.release,
+		TillerNamespace: c.tillerNS,
+		TillerLabel: c.label,
+	}
+
+	v2Releases, _ := utils.ListReleasesWithKubeConfig(listOptions, c.kubeConfig, c.kubeContext)
+	if len(v2Releases) > 0 {
+		log.Printf("A v2 Release of %s was found", c.release)
+		v2ReleaseFound = true
+	} else {
+		log.Printf("There is no v2 Release of %s to migrate", c.release)
+		return nil
+	}
+
+	// Check for existence of v3 Release
+	var v3ReleaseFound bool
+
+	kc := common.KubeConfig{
+		Context: c.kubeContext,
+		File:    c.kubeConfig,
+	}
+
+	v3cfg, err := v3.GetActionConfig(c.namespace, kc)
+	if err != nil {
+		return err
+	}
+
+	_, err = v3cfg.Releases.Deployed(c.release)
+	if err != nil {
+		log.Printf("No v3 Release of %s found", c.release)
+	} else {
+		log.Printf("A v3 Release of %s was found", c.release)
+		v3ReleaseFound = true
+	}
+
+	if v2ReleaseFound && v3ReleaseFound {
+		return fmt.Errorf("Release %s, has entries both in v2 and v3 format", c.release)
+	}
+
+	return c.cmd.Run()
```

Comment:
This isn't really easy to mock out as is, since even if you make some of the discovery logic more testable this will still execute against whatever was setup in the `Prepare` method. I think there should more consistency between the configuration of the `Prepare` and `Execute` methods.

Reply:
Agreed, I'm in the process of doing a refactoring to address these and other things